x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java

```diff
@@ -0,0 +1,121 @@
+import org.elasticsearch.search.aggregations.metrics.MetricsAggregator;
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.sort.BucketedSort;
+import org.elasticsearch.search.sort.SortValue;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Collects the {@code top_metrics} aggregation.
+ *
+ * This extends {@linkplain NumericMetricsAggregator.MultiValue} as a compromise
+ * to allow sorting on the metric. Right now it only collects a single metric
+ * but we expect it to collect a list of them in the future. Also in the future
+ * we expect it to allow collecting non-string metrics which'll change how we
+ * do the inheritance. Finally, we also expect it to allow collecting more than
+ * one document worth of metrics. Once that happens we'll need to come up with
+ * some way to pick which document's metrics to use for the sort.
+ */
+class TopMetricsAggregator extends MetricsAggregator {
+    private final BucketedSort sort;
+    private final String metricName;
+    private final ValuesSource.Numeric metricValueSource;
+    private DoubleArray values;
+
+    TopMetricsAggregator(String name, SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+            Map<String, Object> metaData, BucketedSort sort,
+            String metricName, ValuesSource.Numeric metricValueSource) throws IOException {
+        super(name, context, parent, pipelineAggregators, metaData);
+        this.sort = sort;
+        this.metricName = metricName;
+        this.metricValueSource = metricValueSource;
+        if (metricValueSource != null) {
+            values = context.bigArrays().newDoubleArray(2, false);
+            values.fill(0, values.size(), Double.NaN);
+        }
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        boolean needs = (sort != null && sort.needsScores()) || (metricValueSource != null && metricValueSource.needsScores());
+        return needs ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        assert sub == LeafBucketCollector.NO_OP_COLLECTOR : "Expected noop but was " + sub.toString();
+
+        if (metricValueSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        BucketedSort.Leaf leafSort = sort.forLeaf(ctx);
+        // TODO allow configuration of value mode
+        NumericDoubleValues metricValues = MultiValueMode.AVG.select(metricValueSource.doubleValues(ctx));
+
+        return new LeafBucketCollector() { // TODO do we need to extend *Base*? It doesn't look like we use it.
```

Comment:
Not sure I understand the comment.  Extend "Base"?

Reply:
I believe this is a leftover. I'll remove.